onnx/shape_inference/shape_inference.h

```diff
@@ -77,22 +77,60 @@ void InferShapes(ModelProto& m) {
 
       schema->GetShapeInferenceFunction()(ctx);
 
-      // TODO: this handles the simple case of adding value_infos when
-      // there were none before. It should be enhanced to refine
-      // partially-specified types as well.
-
       for (int i = 0; i < n.output_size(); i++) {
         const auto& output = n.output(i);
-        if (ctx.getOutputType(i)->elem_type() == TensorProto::UNDEFINED) {
+        const auto& inferredType = *ctx.getOutputType(i);
+
+        // In this case, we have no new information, so don't bother
+        // to add a contentless ValueInfo.
+        if (inferredType.elem_type() == TensorProto::UNDEFINED &&
+            inferredType.shape().dim_size() == 0) {
```

Comment:
you need to check that shape() is not set. Size zero shape basically means scalar

Reply:
++